Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,183 @@
+    for snapshot: DocumentSnapshot,
+    buildSettings: SwiftCompileCommand?
+  ) async throws -> RelatedFullDocumentDiagnosticReport {
+    if let reportTask = reportTask(for: snapshot.id, buildSettings: buildSettings) {
+      return try await reportTask.value
+    }
+    let reportTask: Task<RelatedFullDocumentDiagnosticReport, Error>
+    if let buildSettings, !buildSettings.isFallback {
+      reportTask = Task {
+        return try await requestReport(with: snapshot, compilerArgs: buildSettings.compilerArgs)
+      }
+    } else {
+      logger.log(
+        "Producing syntactic diagnostics from the built-in swift-syntax because we \(buildSettings != nil ? "have fallback build settings" : "don't have build settings", privacy: .public))"
+      )
+      // If we don't have build settings or we only have fallback build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      reportTask = Task {
+        return try await requestFallbackReport(with: snapshot)
+      }
+    }
+    setReportTask(for: snapshot.id, buildSettings: buildSettings, reportTask: reportTask)
+    return try await reportTask.value
+  }
+
+  func removeItemsFromCache(with uri: DocumentURI) async {
+    for item in reportTaskCache {
+      if item.snapshotID.uri == uri {
+        item.reportTask.cancel()
+      }
+    }
+    reportTaskCache.removeAll(where: { $0.snapshotID.uri == uri })
+  }
+
+  private func requestReport(
+    with snapshot: DocumentSnapshot,
+    compilerArgs: [String]
+  ) async throws -> LanguageServerProtocol.RelatedFullDocumentDiagnosticReport {
+    try Task.checkCancellation()
+
+    let keys = self.keys
+
+    let skreq = sourcekitd.dictionary([
+      keys.request: requests.diagnostics,
+      keys.sourcefile: snapshot.uri.pseudoPath,
+      keys.compilerargs: compilerArgs as [SKDValue],
+    ])
+
+    let dict = try await self.sourcekitd.send(skreq, fileContents: snapshot.text)
+
+    try Task.checkCancellation()
+    guard (try? documentManager.latestSnapshot(snapshot.uri).id) == snapshot.id else {
+      // Check that the document wasn't modified while we were getting diagnostics. This could happen because we are
+      // calling `fullDocumentDiagnosticReport` from `publishDiagnosticsIfNeeded` outside of `messageHandlingQueue`
+      // and thus a concurrent edit is possible while we are waiting for the sourcekitd request to return a result.
+      throw ResponseError.unknown("Document was modified while loading diagnostics")
+    }
+
+    var diagnostics: [Diagnostic] = []
+    dict[keys.diagnostics]?.forEach { _, diag in
+      if let diag = Diagnostic(
+        diag,
+        in: snapshot,
+        useEducationalNoteAsCode: self.clientHasDiagnosticsCodeDescriptionSupport
+      ) {
+        diagnostics.append(diag)
+      }
+      return true
+    }
```

Comment:
I recently introduced `map` and `compactMap` on `SKDResponseArray`, which we could use here

```suggestion
    let diagnostics: [Diagnostic] = dict[keys.diagnostics]?.compactMap({ diag in
      Diagnostic(
        diag,
        in: snapshot,
        useEducationalNoteAsCode: self.clientHasDiagnosticsCodeDescriptionSupport
      )
    }) ?? []
```